### DISCUSSION TITLE
uv_loop_close() always return BUSY if a timer is inited.

### DISCUSSION THREAD

**Discussion by qishiyexu** [libuv/libuv#3342] [Q&A] [answered]
* **Version**: 1.40
* **Platform**: mac osx

When a timer is inited using uv_timer_init(),  a handle will be pushed into loop->handle_queue; when a timer handle is closed using uv_close(), it's only removed from (handle)->handle_queue) but still exist in loop->handle_queue, causing uv_loop_close return EBUSY.

> mainloop:
uv_timer_start(timer);
uv_run(loop, UV_RUN_DEFAULT);
int result = uv_loop_close(loop);  //return -16
if (result == UV_EBUSY) {
         uv_walk (loop, on_uv_walk, obj);
        uv_run(loop, UV_RUN_DEFAULT);
        result = uv_loop_close(loop);   //still return -16
}

>void on_uv_walk(uv_handle_t* handle, void* arg)
{
    if (!uv_is_closing(handle)) {    //FALSE, handle is closing
        uv_close(handle, on_uv_close);
    }
}

>void close_timer()
{
uv_timer_stop(timer);
uv_close(timer, _close_cb);  //_close_cb has been executed.
}

**Comment by santigimeno** [comment, marked as answer]
This is expected, the [uv_loop_close()](http://docs.libuv.org/en/v1.x/loop.html#c.uv_loop_close) documentation states: 

>  Call this function only when the loop has finished executing and all open handles and requests have been closed, or it will return UV_EBUSY.

So calling `uv_loop_close()` from the `uv_close()` callback should make the trick.

**Reply by qishiyexu** [reply]
But uv_loop_close() would return BUSY even in uv_close() callback, since the timer will never be removed from loop->handle_queue.

**Reply by santigimeno** [reply]
Yes, afaict, it's removed in the [uv_finish_close()](https://github.com/libuv/libuv/blob/v1.x/src/unix/core.c#L304) function.
Btw, calling `uv_loop_close()` from a handle callback is a bad idea. This code works for me:
```c
#include "uv.h"

#include <assert.h>

void on_uv_close(uv_handle_t* handle) {
}

void on_uv_walk(uv_handle_t* handle, void* arg) {
  if (!uv_is_closing(handle)) { //FALSE, handle is closing
    uv_close(handle, on_uv_close);
  }
}

int main() {
  uv_timer_t timer;
  int r;
  r = uv_timer_init(uv_default_loop(), &timer);
  assert(r == 0);
  r = uv_run(uv_default_loop(), UV_RUN_DEFAULT);
  assert(r == 0);
  // It returns busy cause the timer handle is not closed.
  r = uv_loop_close(uv_default_loop());
  assert(r == UV_EBUSY);
  // Close pending handles
  uv_walk(uv_default_loop(), on_uv_walk, NULL);
  // run the loop until there are no pending callbacks
  do {
    r = uv_run(uv_default_loop(), UV_RUN_ONCE);
  } while(r != 0);
  // Now we're safe.
  r = uv_loop_close(uv_default_loop());
  assert(r == 0);
}
```

**Reply by qishiyexu** [reply]
You are right. Thanks for your answer.
